tt_metal/impl/dispatch/arch.cpp

```diff
@@ -0,0 +1,738 @@
+    for (auto fd_kernel : node_id_to_kernel) {
+        if (auto router_kernel = dynamic_cast<EthRouterKernel *>(fd_kernel)) {
+            chip_id_t router_device_id = router_kernel->GetDeviceId();
+            chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(router_device_id);
+
+            // Router placement CQID already set for mmio device above, just need to do this for remotes
+            if (router_device_id != mmio_device_id) {
+                router_kernel->SetPlacementCQID(device_id_to_num_routers[router_device_id]);
+            }
+            device_id_to_remaining_routers[router_device_id]++;
+            device_id_to_num_routers[router_device_id]++;
+        }
+    }
+    for (auto fd_kernel : node_id_to_kernel) {
+        uint32_t tunnel_stop = device_id_to_tunnel_stop[fd_kernel->GetDeviceId()];
+        if (auto tunneler_kernel = dynamic_cast<EthTunnelerKernel *>(fd_kernel)) {
+            // Local tunneler needs to match remote tunneler on previous chip in the tunnel
+            if (!tunneler_kernel->IsRemote()) {
+                TT_ASSERT(tunnel_stop != 0);
+                tunnel_stop--;
+            }
+            // # of VCs is return VC + total VCs for tunnel (num_hw_cqs per remote) - num_hw_cqs VCs per remote stop
+            tunneler_kernel->SetVCCount(1 + (tunnel_depth - 1) * num_hw_cqs - tunnel_stop * num_hw_cqs);
+        } else if (auto router_kernel = dynamic_cast<EthRouterKernel *>(fd_kernel)) {
+            // Router has the same VCs as the remote tunneler for MMIO, same as local tunneler for remote
+            if (tunnel_stop != 0) {
+                tunnel_stop--;
+            }
+            uint32_t router_vcs_for_device = (tunnel_depth - 1) * num_hw_cqs - tunnel_stop * num_hw_cqs;
+            // Divide the VCs between routers on one device
+            uint32_t remaining_routers = device_id_to_remaining_routers[router_kernel->GetDeviceId()];
+            uint32_t num_routers = device_id_to_num_routers[router_kernel->GetDeviceId()];
+
+            // Special case for MMIO chips, can have routers servicing different tunnels
+            chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(router_kernel->GetDeviceId());
+            if (router_kernel->GetDeviceId() == mmio_device_id) {
+                uint32_t num_tunnels = tt::Cluster::instance().get_tunnels_from_mmio_device(mmio_device_id).size();
+                num_routers /= num_tunnels;
+                uint32_t router_vcs =  (router_vcs_for_device + num_routers - 1) / num_routers;
+                router_kernel->SetVCCount(router_vcs);
+            } else {
+                uint32_t router_vcs = (router_vcs_for_device + remaining_routers - 1) / num_routers;
+                router_kernel->SetVCCount(router_vcs);
+            }
+            device_id_to_remaining_routers[router_kernel->GetDeviceId()]--;
+        }
+    }
+}
+
+std::unique_ptr<Program> create_cq_program(Device *device) {
```

Comment:
```suggestion
std::unique_ptr<Program> compile_cq_program(Device *device) {
```

Reply:
This function does also create a `Program`, will rename it to `create_and_compile_cq_program`